src/slurm_plugin/clustermgtd.py

```diff
@@ -964,6 +989,93 @@ def _is_node_in_replacement_valid(self, node, check_node_is_valid):
+        self._set_ice_compute_resources_to_down(active_nodes)
+
+    def _update_insufficient_capacity_compute_resources(self, ice_compute_resources_and_nodes_map: dict):
+        """Add compute resource to insufficient_capacity_compute_resources if node is ICE node."""
+        for compute_resource, data in ice_compute_resources_and_nodes_map.items():
+            if compute_resource not in self._insufficient_capacity_compute_resources:
+                self._insufficient_capacity_compute_resources[compute_resource] = ComputeResourceFailureEvent(
+                    self._current_time, data.error_code
+                )
+
+    def _reset_timeout_expire_compute_resources(self, ice_compute_resources_and_nodes_map: dict):
+        """Reset compute resources which insufficient_capacity_timeout expire."""
+        # Find insufficient_capacity_timeout compute resources
+        if not self._insufficient_capacity_compute_resources:
+            return
+        log.info(
+            "The following compute resources are in down state due to insufficient capacity: %s, "
+            "compute resources will be reset after insufficient capacity timeout (%s seconds) expire",
+            self._insufficient_capacity_compute_resources,
+            self._config.insufficient_capacity_timeout,
+        )
+        timeout_expired_compute_resources = self._find_insufficient_capacity_timeout_expire_compute_resources()
+
+        # Reset nodes which insufficient capacity timeout expire
+        if timeout_expired_compute_resources:
+            self._reset_insufficient_capacity_timeout_expire_nodes(
+                timeout_expired_compute_resources, ice_compute_resources_and_nodes_map
+            )
+
+    def _set_ice_compute_resources_to_down(self, active_nodes: list):
+        """Set powered_down nodes which belong to insufficient capacity compute resources to down."""
+        if not self._insufficient_capacity_compute_resources:
+            return
+        nodes_to_down = {}
+        for node in active_nodes:
+            compute_resource_name = node.get_compute_resource_name()
+            if compute_resource_name in self._insufficient_capacity_compute_resources:
+                if node.is_powered_down_and_not_ice():
+                    error_code = self._insufficient_capacity_compute_resources.get(compute_resource_name).error_code
+                    nodes_to_down.setdefault(error_code, []).append(node.name)
+        if nodes_to_down:
+            for error_code, node_list in nodes_to_down.items():
+                log.info(
+                    "Setting following nodes into DOWN state due to insufficient capacity: %s",
+                    print_with_count(node_list),
+                )
+                set_nodes_down(
+                    node_list, reason=f"(Code:{error_code})Temporarily disabling node due to insufficient capacity"
+                )
+
+    def _find_insufficient_capacity_timeout_expire_compute_resources(self) -> dict:
```

Comment:
Can we simplify this and have it return a set of CR names?

Reply:
Yes, only the cr names are used here. Changed